app/src/main/java/space/taran/arknavigator/mvp/model/repo/tags/TagsStorageRepo.kt

```diff
@@ -11,29 +13,31 @@ class TagsStorageRepo(
     private val foldersRepo: FoldersRepo,
     private val indexRepo: ResourcesIndexRepo
 ) {
+    private val provideMutex = Mutex()
     private val storageByRoot = mutableMapOf<Path, PlainTagsStorage>()
 
     suspend fun provide(rootAndFav: RootAndFav): TagsStorage = withContext(Dispatchers.IO) {
-        val allRoots = foldersRepo.query().succeeded.keys
-        val roots = if (rootAndFav.isAllRoots()) allRoots else listOf(rootAndFav.root!!)
+        provideMutex.withLock {
+            val roots = foldersRepo.resolveRoots(rootAndFav)
```

Comment:
We can do locking after resolving roots, the same as in `ResourcesIndexRepo.provide()`.